test/test-executor.c

```diff
@@ -0,0 +1,322 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "task.h"
+
+static int work_cb_count;
+static int after_work_cb_count;
+static uv_work_t work_req;
+static char data;
+
+
+static void work_cb(uv_work_t* req) {
+  ASSERT(req == &work_req);
+  ASSERT(req->data == &data);
+  work_cb_count++;
+}
+
+
+static void after_work_cb(uv_work_t* req, int status) {
+  ASSERT(status == 0);
+  ASSERT(req == &work_req);
+  ASSERT(req->data == &data);
+  after_work_cb_count++;
+}
+
+
+TEST_IMPL(executor_queue_work_simple) {
+  int r;
+
+  work_req.data = &data;
+  r = uv_queue_work(uv_default_loop(), &work_req, work_cb, after_work_cb);
+  ASSERT(r == 0);
+  uv_run(uv_default_loop(), UV_RUN_DEFAULT);
+
+  ASSERT(work_cb_count == 1);
+  ASSERT(after_work_cb_count == 1);
+
+  MAKE_VALGRIND_HAPPY();
+  return 0;
+}
+
+
+TEST_IMPL(executor_queue_work_einval) {
+  int r;
+
+  work_req.data = &data;
+  r = uv_queue_work(uv_default_loop(), &work_req, NULL, after_work_cb);
+  ASSERT(r == UV_EINVAL);
+
+  uv_run(uv_default_loop(), UV_RUN_DEFAULT);
+
+  ASSERT(work_cb_count == 0);
+  ASSERT(after_work_cb_count == 0);
+
+  MAKE_VALGRIND_HAPPY();
+  return 0;
+}
+
+/* Define an toy_executor.
+ * This is a trivial one-thread producer-consumer setup with a fixed buffer of work. */
+#define TOY_EXECUTOR_MAX_REQUESTS 100
+static uv_executor_t toy_executor;
+static struct toy_executor_data {
+  /* Do not hold while executing work. */
+  uv_mutex_t mutex;
+
+  int times_submit_called;
+  int times_cancel_called;
+
+  int n_completed;
+
+  int head;
+  int tail;
+  /* Queue with space for some extras */
+  uv_work_t* queued_work[TOY_EXECUTOR_MAX_REQUESTS + 10];
+
+  int no_more_work_coming;
+  uv_sem_t thread_exiting;
+
+  uv_thread_t thread;
+  uv_executor_t *executor;
+} toy_executor_data;
+
+static void worker(void* arg) {
+  struct toy_executor_data* data;
+  uv_work_t* work;
+  int should_break;
+
+  data = (struct toy_executor_data *) arg;
+
+  should_break = 0;
+  while (1) {
+    /* Run until we (1) have no work, and (2) see data->no_more_work_coming. */
+
+    /* Run pending work. */
+    while (1) {
+      uv_mutex_lock(&data->mutex);
+      if (data->head == data->tail) {
+        uv_mutex_unlock(&data->mutex);
+        break;
+      }
+
+      printf("worker: found %d work -- head %d tail %d\n", data->tail - data->head, data->head, data->tail);
```

Comment:
Long line.

Reply:
This will be cut when I purge the debugging messages.